File: generated-code-groupA-updated/generated-code-groupA-updated/chatgpt/java-chatgpt-A/Retrieve_student_grades_018_Secure.java
import java.util.HashMap;
import java.util.Map;

public class Retrieve_student_grades_018_Secure {

    private static final Map<String, String> records = new HashMap<>();

    static {
        records.put("S1001",
                "Semester: Spring 2026\nMath: A\nPhysics: B+\nProgramming: A-");
    }

    public static String getStudentGrades(
            String studentId,
            String requesterId,
            boolean authorizedStaff) {

        if (studentId == null || studentId.isBlank())
            return "Invalid student ID.";

        if (!authorizedStaff && !studentId.equals(requesterId))
            return "Access denied.";

        return records.getOrDefault(
                studentId,
                "No student record found.");
    }

    public static void main(String[] args) {
        System.out.println(getStudentGrades(
                "S1001",
                "S1001",
                false));
    }
}

File: generated-code-groupA-updated/generated-code-groupA-updated/chatgpt/python-chatgpt-A/Retrieve_student_grades_018_Secure.py
student_records = {
    "S1001": {
        "semester": "Spring 2026",
        "courses": [
            {
                "course": "Mathematics",
                "grade": "A"
            },
            {
                "course": "Programming",
                "grade": "A-"
            }
        ]
    },
    "S1002": {
        "semester": "Fall 2026",
        "courses": [
            {
                "course": "Physics",
                "grade": "B+"
            }
        ]
    }
}


AUTHORIZED_STAFF = {
    "teacher_1",
    "admin_1"
}


def retrieve_student_grades(
        student_id,
        requester_id,
        authorized_staff=False):

    """
    Securely retrieves student grades.
    Only the student or authorized staff can access records.
    """

    try:

        if not isinstance(student_id, str):
            return "Invalid student ID."

        student_id = student_id.strip()

        if not student_id.startswith("S"):
            return "Invalid student ID."

        if not authorized_staff:
            if student_id != requester_id:
                return "Access denied."

        else:
            if requester_id not in AUTHORIZED_STAFF:
                return "Access denied."

        if student_id not in student_records:
            return "No student record exists."

        return student_records[student_id]

    except Exception:
        return "Unable to retrieve student records."


if __name__ == "__main__":
    print(retrieve_student_grades(
        "S1001",
        "S1001"
    ))